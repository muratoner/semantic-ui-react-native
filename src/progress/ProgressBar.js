import React, { useMemo } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { ProgressBarProps } from '../';
import { UtilColor } from '../utils';

export default (props: ProgressBarProps) => {
	let {
		style,
		progress,
		height,
		showProgressValue,
		bottomText,
		bottomTextIncludeProgress,
		autoColored
	} = props;

	const autoColorValue = useMemo(() => {
		var red = 0,
			green = 0;
		if (progress >= 50) {
			red = 255 - Math.round(((progress - 50) / 50) * 255);
			green = 255;
		} else {
			red = 255;
			green = Math.round((progress / 50) * 255);
		}
		return `rgb(${red}, ${green}, 0)`;
	}, [progress]);

	const autoBgColor = useMemo(() => {
		if (autoColored) {
			return {
				backgroundColor: autoColorValue
			};
		}
		return null;
	}, [progress, autoColorValue]);

	const autoProgressTextColor = useMemo(() => {
		if (autoColored) {
			return {
				color: UtilColor.contrast(autoColorValue)
			};
		}
		return null;
	}, [autoBgColor]);

	const autoBottomTextColor = useMemo(() => {
		if (autoColored) {
			return {
				color: autoColorValue
			};
		}
		return null;
	}, [autoBgColor]);

	const computedStyle = useMemo(() => {
		return { height: height || 20 };
	}, [height]);

	return (
		<>
			<View style={[styles.container, style]}>
				<View
					style={[
						styles.fill,
						computedStyle,
						{ width: `${showProgressValue && progress < 10 ? 10 : progress}%` },
						autoBgColor
					]}
				>
					{showProgressValue && (
						<Text
							numberOfLines={1}
							adjustsFontSizeToFit
							style={StyleSheet.flatten([
								styles.progressBarText,
								autoProgressTextColor
							])}
						>
							{progress}%
						</Text>
					)}
				</View>
			</View>
			{bottomText && (
				<Text
					style={StyleSheet.flatten([styles.bottomText, autoBottomTextColor])}
				>
					{bottomTextIncludeProgress ? `${progress}% ` : ''}
					{bottomText}
				</Text>
			)}
		</>
	);
};

const styles = StyleSheet.create({
	bottomText: {
		color: '#000',
		marginTop: 5,
		alignSelf: 'center',
		fontWeight: '500'
	},
	progressBarText: {
		color: '#fff',
		paddingHorizontal: 5,
		alignSelf: 'flex-end'
	},
	fill: {
		backgroundColor: '#888',
		borderRadius: 10,
		justifyContent: 'center'
	},
	container: {
		overflow: 'hidden',
		borderRadius: 10,
		backgroundColor: 'rgba(0,0,0,.1)'
	}
});
