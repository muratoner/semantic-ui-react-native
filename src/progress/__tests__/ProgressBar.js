import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ProgressBar from '../ProgressBar';

describe('ProgressBar Component', () => {
	it('should render without issues', () => {
		const component = shallow(<ProgressBar />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with 0 progress', () => {
		const component = shallow(<ProgressBar progress={0} />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with 90 progress', () => {
		const component = shallow(<ProgressBar progress={60} />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with 100 progress', () => {
		const component = shallow(<ProgressBar progress={100} />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with 100 progress and custom style', () => {
		const component = shallow(
			<ProgressBar style={{ backgroundColor: '#f90809' }} progress={100} />
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});
});
