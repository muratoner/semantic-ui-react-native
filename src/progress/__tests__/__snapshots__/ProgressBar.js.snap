// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar Component should render with 0 progress 1`] = `
<Fragment>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0,0,0,.1)",
          "borderRadius": 10,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#888",
            "borderRadius": 10,
            "justifyContent": "center",
          },
          Object {
            "height": 20,
          },
          Object {
            "width": "0%",
          },
          null,
        ]
      }
    />
  </View>
</Fragment>
`;

exports[`ProgressBar Component should render with 90 progress 1`] = `
<Fragment>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0,0,0,.1)",
          "borderRadius": 10,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#888",
            "borderRadius": 10,
            "justifyContent": "center",
          },
          Object {
            "height": 20,
          },
          Object {
            "width": "30%",
          },
          null,
        ]
      }
    />
  </View>
</Fragment>
`;

exports[`ProgressBar Component should render with 100 progress 1`] = `
<Fragment>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0,0,0,.1)",
          "borderRadius": 10,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#888",
            "borderRadius": 10,
            "justifyContent": "center",
          },
          Object {
            "height": 20,
          },
          Object {
            "width": "100%",
          },
          null,
        ]
      }
    />
  </View>
</Fragment>
`;

exports[`ProgressBar Component should render with 100 progress and custom style 1`] = `
<Fragment>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0,0,0,.1)",
          "borderRadius": 10,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#f90809",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#888",
            "borderRadius": 10,
            "justifyContent": "center",
          },
          Object {
            "height": 20,
          },
          Object {
            "width": "100%",
          },
          null,
        ]
      }
    />
  </View>
</Fragment>
`;

exports[`ProgressBar Component should render with auto colored 1`] = `
<Fragment>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0,0,0,.1)",
          "borderRadius": 10,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#888",
            "borderRadius": 10,
            "justifyContent": "center",
          },
          Object {
            "height": 20,
          },
          Object {
            "width": "70%",
          },
          Object {
            "backgroundColor": "rgb(153, 255, 0)",
          },
        ]
      }
    />
  </View>
</Fragment>
`;

exports[`ProgressBar Component should render without issues 1`] = `
<Fragment>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0,0,0,.1)",
          "borderRadius": 10,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#888",
            "borderRadius": 10,
            "justifyContent": "center",
          },
          Object {
            "height": 20,
          },
          Object {
            "width": "undefined%",
          },
          null,
        ]
      }
    />
  </View>
</Fragment>
`;
